File: android/beagle/src/test/java/br/com/zup/beagle/android/logger/BeagleLoggerProxyTest.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.android.logger

import br.com.zup.beagle.android.extensions.once
import br.com.zup.beagle.android.setup.BeagleEnvironment
import br.com.zup.beagle.android.testutil.RandomData
import io.mockk.*
import kotlin.math.log
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test

private val LOG = RandomData.string()

class BeagleLoggerProxyTest {

    private val logger = mockk<BeagleLogger>(relaxUnitFun = true, relaxed = true)

    @BeforeEach
    fun setUp() {
        mockkObject(BeagleEnvironment)

        every { BeagleEnvironment.beagleSdk.config.isLoggingEnabled } returns true
        every { BeagleEnvironment.beagleSdk.logger } returns logger
        BeagleLoggerProxy.logger = logger
    }

    @AfterEach
    fun tearDown() {
        unmockkAll()
    }

    @Test
    fun warning_should_call_Log_w_if_is_enable() {
        // When
        BeagleLoggerProxy.warning(LOG)

        // Then
        verify(exactly = once()) { logger.warning(LOG) }
    }

    @Test
    fun warning_should_not_call_Log_w_if_is_not_enable() {
        // Given
        every { BeagleEnvironment.beagleSdk.config.isLoggingEnabled } returns false

        // When
        BeagleLoggerProxy.warning(LOG)

        // Then
        verify(exactly = 0) { logger.warning(LOG) }
    }

    @Test
    fun error_should_call_Log_w_if_is_enable() {

        // When
        BeagleLoggerProxy.error(LOG)

        // Then
        verify(exactly = once()) { logger.error(LOG) }
    }

    @Test
    fun error_should_not_call_Log_w_if_is_not_enable() {
        // Given
        every { BeagleEnvironment.beagleSdk.config.isLoggingEnabled } returns false

        // When
        BeagleLoggerProxy.error(LOG)

        // Then
        verify(exactly = 0) { logger.error(LOG) }
    }

    @Test
    fun info_should_call_Log_w_if_is_enable() {
        // When
        BeagleLoggerProxy.info(LOG)

        // Then
        verify(exactly = once()) { logger.info(LOG) }
    }

    @Test
    fun info_should_not_call_Log_w_if_is_not_enable() {
        // Given
        every { BeagleEnvironment.beagleSdk.config.isLoggingEnabled } returns false

        // When
        BeagleLoggerProxy.info(LOG)

        // Then
        verify(exactly = 0) { logger.info(LOG) }
    }
}

File: flutter/beagle/test/logger/beagle_logger_proxy_test.dart
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import 'package:beagle/logger/beagle_logger.dart';
import 'package:beagle/logger/beagle_logger_proxy.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mockito/mockito.dart';

void main() {
  final logger = BeagleLoggerMock();
  const logMessage = 'log message';
  final logException = Exception('log exception');

  group('Given a BeagleLoggerProxy with a non null logger', () {
    final beagleLoggerProxy = BeagleLoggerProxy(logger: logger);

    group('When I call error method', () {
      test('Then it should call logger error method', () {
        beagleLoggerProxy.error(logMessage);
        verify(logger.error(logMessage)).called(1);
      });
    });

    group('When I call errorWithException method', () {
      test('Then it should call logger errorWithException method', () {
        beagleLoggerProxy.errorWithException(logMessage, logException);
        verify(logger.errorWithException(logMessage, logException)).called(1);
      });
    });

    group('When I call info method', () {
      test('Then it should call logger info method', () {
        beagleLoggerProxy.info(logMessage);
        verify(logger.info(logMessage)).called(1);
      });
    });

    group('When I call warning method', () {
      test('Then it should call logger warning method', () {
        beagleLoggerProxy.warning(logMessage);
        verify(logger.warning(logMessage)).called(1);
      });
    });
  });

  group('Given a BeagleLoggerProxy with a null logger', () {
    final beagleLoggerProxy = BeagleLoggerProxy();

    group('When I call error method', () {
      test('Then it should not throw any exception', () {
        beagleLoggerProxy.error(logMessage);
      });
    });

    group('When I call errorWithException method', () {
      test('Then it should not throw any exception', () {
        beagleLoggerProxy.errorWithException(logMessage, logException);
      });
    });

    group('When I call info method', () {
      test('Then it should not throw any exception', () {
        beagleLoggerProxy.info(logMessage);
      });
    });

    group('When I call warning method', () {
      test('Then it should not throw any exception', () {
        beagleLoggerProxy.warning(logMessage);
      });
    });
  });
}

class BeagleLoggerMock extends Mock implements BeagleLogger {}
